File: tsp.cpp
#define _CRT_SECURE_NO_WARNINGS

#include <iostream>
#include <cstdint>
#include <functional>
#include <vector>
#include <queue>
#include <stack>
#include <utility>
using namespace std;

const size_t N = 50;
const uint32_t INF = uint32_t(1e+9);

typedef pair < size_t, size_t > Edge;
Edge NullEdge(N, N);

enum class EdgeType
{
	Outgoing,
	Incoming
};

struct PartialSolution
{
	bool operator>(const PartialSolution& other) const
	{
		return LowerBoundTimesTwo > other.LowerBoundTimesTwo;
	}

	PartialSolution WithEdge(Edge pivot, uint32_t D[N][N])
	{
		auto i = pivot.first, j = pivot.second;
		
		PartialSolution child = *this;
		child.Cost += D[i][j];
		for (size_t k = 0; k < n; k++)
		{
			child.Constraints[i][k] = child.Constraints[k][j] = -1;
			child.Reduced[i][k] = child.Reduced[k][j] = INF;
		}
		child.EnabledEdges++;
		child.Constraints[i][j] = 1;
		child.Constraints[j][i] = -1;

		auto subpathTo = child.TraverseSubPath(i, EdgeType::Outgoing);
		auto subpathFrom = child.TraverseSubPath(i, EdgeType::Incoming);
		if (subpathTo.size() + subpathFrom.size() - 1 != n)
		{
			child.Constraints[subpathTo.back()][subpathFrom.back()] = -1;
			child.Reduced[subpathTo.back()][subpathFrom.back()] = INF;
		}

		child.Reduce();
		return child;
	}

	PartialSolution WithoutEdge(Edge pivot, uint32_t D[N][N])
	{
		auto i = pivot.first, j = pivot.second;
		
		PartialSolution child = *this;
		child.DisabledEdges++;
		child.Constraints[i][j] = -1;
		child.Reduced[i][j] = INF;
		child.Reduce(EdgeType::Outgoing, i);
		child.Reduce(EdgeType::Incoming, j);

		return child;
	}

	Edge ChoosePivotEdge()
	{
		auto minStride = [&](size_t except, size_t k, size_t kStride) {uint32_t m = INF; for (size_t i = 0; i < n; i++) if(i != except) m = min(m, *(&Reduced[0][0] + IK(i, k, kStride))); return m;  };
		auto rowMin = [&](size_t k, size_t except) {return minStride(except, k, N); };
		auto columnMin = [&](size_t k, size_t except) {return minStride(except, k, 1); };
		
		uint32_t bestIncrease = 0;
		Edge bestPivot = NullEdge;
		for (size_t i = 0; i < n; i++)
		{
			for (size_t j = 0; j < n; j++)
			{
				if (Constraints[i][j] == 0 && Reduced[i][j] == 0)
				{
					auto increase = rowMin(i, j) + columnMin(j, i);
					if (increase > bestIncrease)
					{
						bestIncrease = increase;
						bestPivot = Edge(i, j);
					}
				}
			}
		}

		return bestPivot;
	}

	vector<size_t> TraverseSubPath(size_t cur, EdgeType edgeType)
	{
		auto stride = edgeType == EdgeType::Outgoing ? 1 : N;
		vector<size_t> subpath{ cur };
		for (size_t k = 0; k < n; k++)
		{
			size_t next = N;
			for (size_t i = 0; i < n; i++)
			{
				if (*(&Constraints[0][0] + IK(cur, i, stride)) == 1)
				{
					next = i;
					break;
				}
			}

			if (next == N)
				break;

			subpath.push_back(next);
			cur = next;
		}
		return subpath;
	}

	void Reduce(EdgeType edgeType, size_t i)
	{
		auto kStride = edgeType == EdgeType::Outgoing ? 1 : N;
		
		uint32_t m = INF;
		for (size_t k = 0; k < n; k++)
			if (*(&Constraints[0][0] + IK(i, k, kStride)) != -1)
				m = min(m, *(&Reduced[0][0] + IK(i, k, kStride)));
		
		if (m != INF)
		{
			for (size_t k = 0; k < n; k++)
				*(&Reduced[0][0] + IK(i, k, kStride)) -= m;
			LowerBoundTimesTwo += m;
		}
	}

	void Reduce()
	{
		for (size_t i = 0; i < n; i++)
			Reduce(EdgeType::Outgoing, i);

		for (size_t j = 0; j < n; j++)
			Reduce(EdgeType::Incoming, j);
	}

	void Print(uint32_t D[N][N])
	{
		std::cout << Cost - D[n - 1][0] << std::endl;
		//for (size_t i = 1; i < n; i++)
		//	std::cout << Path[i - 1] << " " << Path[i] << " " << D[Path[i - 1]][Path[i]] << std::endl;
	}

	bool IsComplete()
	{
		Path = TraverseSubPath(0, EdgeType::Outgoing);
		return Path.size() == n + 1 && Path[n - 1] == n - 1;
	}

	int IK(size_t i, size_t k, size_t kStride)
	{
		return (N + 1 - kStride)*i + kStride*k;
	}

	PartialSolution(size_t n, uint32_t D[N][N]) : n(n)
	{
		memcpy(Reduced, D, sizeof(Reduced[0][0]) * N * N);
		memset(Constraints, 0, sizeof(Constraints[0][0]) * N * N);
		for (size_t i = 0; i < n; i++)
		{
			Reduced[i][i] = INF;
			Constraints[i][i] = -1;
		}
		Cost = 0;

		Reduce();
	}

	PartialSolution()
	{
	}

	size_t EnabledEdges = 0, DisabledEdges = 0;
	size_t n = 0;
	uint32_t Cost = INF;
	uint32_t LowerBoundTimesTwo = 0;
	uint32_t Reduced[N][N];
	int8_t Constraints[N][N];
	vector<size_t> Path;
};

void branch_and_bound(size_t n, uint32_t D[N][N])
{
	PartialSolution bestCompleteSolution;
	PartialSolution root = PartialSolution(n, D).WithEdge(Edge(n - 1, 0), D);

	static priority_queue<PartialSolution, vector<PartialSolution>, greater<PartialSolution> > right;
	static stack<PartialSolution> left;
	left.push(root);

	while (!left.empty() || !right.empty())
	{
		auto currentSolution = !left.empty() ? left.top() : right.top();
		if (!left.empty())
			left.pop();
		else
			right.pop();
	
		if (currentSolution.IsComplete())
		{
			if (currentSolution.Cost < bestCompleteSolution.Cost)
			{
				bestCompleteSolution = currentSolution;
				bestCompleteSolution.Print(D);
			}
		}
		else if (currentSolution.LowerBoundTimesTwo < bestCompleteSolution.Cost)
		{
			auto pivot = currentSolution.ChoosePivotEdge();
			if (pivot != NullEdge)
			{
				auto withPivot = currentSolution.WithEdge(pivot, D);
				auto withoutPivot = currentSolution.WithoutEdge(pivot, D);

				if (withPivot.LowerBoundTimesTwo < bestCompleteSolution.Cost)
					left.push(withPivot);

				if (withoutPivot.LowerBoundTimesTwo < bestCompleteSolution.Cost)
					right.push(withoutPivot);
			}
		}
	}
}

int main(int argc, char* argv[])
{
	freopen(argv[1], "r", stdin);
	
	size_t n;
	uint32_t D[N][N];

	std::cin >> n;
	for (size_t i = 0; i < n; i++)
		for (size_t j = 0; j < n; j++)
			std::cin >> D[i][j];

	branch_and_bound(n, D);
}

File: tsp.rs
use std::env;
use std::cmp::Ordering;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::collections::BinaryHeap;
use std::cmp::min;

const N : usize = 50;
const INF : u32 = 1e+9 as u32;

#[derive(Copy, Clone, Eq, PartialEq)]
struct Edge
{
	first : usize,
	second : usize
}

const NullEdge : Edge = Edge { first: N, second: N};

//TODO: should be a function/closure that does (arr, i, j) => arr[i][j] for Row and arr[j][i] for column
struct RowColumnAccessor;
const Row : RowColumnAccessor = RowColumnAccessor;
const Column : RowColumnAccessor = RowColumnAccessor;

#[derive(Eq, PartialEq)]
struct PartialSolution
{
	n : usize,
	Cost : u32,
	LowerBoundTimesTwo : u32,
	Reduced : Vec<Vec<u32>>,
	Constraints : Vec<Vec<i8>>,
	Path : Vec<usize>
}

impl Ord for PartialSolution
{
	fn cmp(&self, other: &PartialSolution) -> Ordering
	{
		self.LowerBoundTimesTwo.cmp(&other.LowerBoundTimesTwo)
	}
}

impl PartialOrd for PartialSolution
{
	fn partial_cmp(&self, other: &PartialSolution) -> Option<Ordering>
	{
		Some(self.cmp(other))
	}
}

impl Clone for PartialSolution
{
	//TODO: implement proper clone with cloning Reduced and Constraints
	fn clone(&self) -> PartialSolution
	{
		PartialSolution::InfiniteCostSolution()
	}
}

impl PartialSolution
{
	fn ReduceByType(&mut self, a : RowColumnAccessor, i : usize)
	{
		let m = INF;
		for k in 0..self.n
		{
//			if a(self.Constraints, i, k) != -1
			{
//				m = min(m, a(self.Reduced, i, k));
			}
		}
		
		if (m != INF)
		{
			for k in 0..self.n
			{
//				a(self.Reduced, i, k) -= m;
			}
			self.LowerBoundTimesTwo += m;
		}
	}

	fn Reduce(&mut self)
	{
		for i in 0..self.n
		{
			self.ReduceByType(Row, i);
		}

		for j in 0..self.n
		{
			self.ReduceByType(Column, j);
		}
	}

	fn InfiniteCostSolution() -> PartialSolution
	{
		PartialSolution {n : 0, Cost : INF, LowerBoundTimesTwo : 0, Path : vec![], Reduced : vec![vec![0u32; N]; N], Constraints : vec![vec![0i8; N]; N] }
	}

	fn new(n : usize, D : &Vec<Vec<u32>>) -> PartialSolution
	{
		let mut res = PartialSolution {n : n, Cost : 0, LowerBoundTimesTwo : 0, Path : vec![], Constraints : vec![vec![0i8; N]; N], Reduced : D.clone()};

		for i in 0..n
		{
			res.Reduced[i][i] = INF;
			res.Constraints[i][i] = -1;
		}
		
		res.Reduce();
		res
	}
	
	fn ChoosePivotEdge(&self) -> Edge
	{
		//let minRC = |except: usize, k : usize, a : RowColumnAccessor| {let mut m = INF; for i in 0..self.n { if i != except {m = min(m, a(self.Reduced, i, k));} } m };
		//let rowMin = |k : usize, except : usize| minRC(except, k, Row);
		//let columnMin = |k : usize, except : usize| minRC(except, k, Column);
		
		let mut bestIncrease = 0u32;
		let mut bestPivot = NullEdge;
		for i in 0..self.n
		{
			for j in 0..self.n
			{
				if self.Constraints[i][j] == 0 && self.Reduced[i][j] == 0
				{
					let increase = 0u32;//rowMin(i, j) + columnMin(j, i);
					if increase > bestIncrease
					{
						bestIncrease = increase;
						bestPivot = Edge { first : i, second : j};
					}
				}
			}
		}

		return bestPivot;
	}

	fn WithEdge(&self, pivot : Edge, D: &Vec<Vec<u32>>) -> PartialSolution
	{
		let Edge { first: i, second: j } = pivot;
		
		let mut child = self.clone();
		child.Cost += D[i][j];
		for k in 0..self.n
		{
			child.Constraints[i][k] = -1;
			child.Constraints[k][j] = -1;
			child.Reduced[i][k] = INF;
			child.Reduced[k][j] = INF;
		}

		child.Constraints[i][j] = 1;
		child.Constraints[j][i] = -1;

		let subpathTo = child.TraverseSubPath(i, Row);
		let subpathFrom = child.TraverseSubPath(i, Column);
		if subpathTo.len() + subpathFrom.len() - 1 != self.n
		{
			let toBack = *subpathTo.last().unwrap();
			let fromBack = *subpathFrom.last().unwrap(); 		
			child.Constraints[toBack][fromBack] = -1;
			child.Reduced[toBack][fromBack] = INF;
		}

		child.Reduce();
		child
	}

	fn WithoutEdge(&self, pivot : Edge, D: &Vec<Vec<u32>>) -> PartialSolution
	{
		let Edge { first: i, second: j } = pivot;
		
		let mut child = self.clone();
		child.Constraints[i][j] = -1;
		child.Reduced[i][j] = INF;
		child.ReduceByType(Row, i);
		child.ReduceByType(Column, j);

		return child;
	}

	fn TraverseSubPath(&self, mut cur : usize, a : RowColumnAccessor) -> Vec<usize>
	{
		let mut subpath = vec![cur];
		for k in 0..self.n
		{
			let mut next = N;
			for i in 0..self.n
			{
//				if a(Constraints, cur, i)
				{
					next = i;
					break;
				}
			}

			if next == N
			{
				break;
			}

			subpath.push(next);
			cur = next;
		}
		subpath
	}

	fn IsComplete(&mut self) -> bool
	{
		self.Path = self.TraverseSubPath(0, Row);
		self.Path.len() == self.n + 1 && self.Path[self.n - 1] == self.n - 1
	}

	fn Print(&self, D: &Vec<Vec<u32>>)
	{
		println!("{}", self.Cost - D[self.n - 1][0]);
		for i in 1..self.n
		{
			println!("{} {} {}\n", self.Path[i - 1], self.Path[i], D[self.Path[i - 1]][self.Path[i]])
		}
		println!("\n")
	}
}

fn branch_and_bound(n : usize, D : &Vec<Vec<u32>>)
{
	let mut bestCompleteSolution = PartialSolution::InfiniteCostSolution();
	let root = PartialSolution::new(n, D).WithEdge(Edge { first : n - 1, second : 0}, D);

	let mut Q = BinaryHeap::new();
	Q.push(root);

	while let Some(mut currentSolution) = Q.pop()
	{
		//TODO: possibly refactor to use Option, remove Path field from PartialSolution, make Print a free function instead of a method
		if currentSolution.IsComplete()
		{
			if currentSolution.Cost < bestCompleteSolution.Cost
			{
				bestCompleteSolution = currentSolution;
				bestCompleteSolution.Print(D);
			}
		}
		else if currentSolution.LowerBoundTimesTwo < 2 * bestCompleteSolution.Cost
		{
			let pivot = currentSolution.ChoosePivotEdge();
			if pivot != NullEdge
			{
				let withPivot = currentSolution.WithEdge(pivot, D);
				let withoutPivot = currentSolution.WithoutEdge(pivot, D);

				if withPivot.LowerBoundTimesTwo < 2 * bestCompleteSolution.Cost
				{
					Q.push(withPivot);
				}

				if withoutPivot.LowerBoundTimesTwo < 2 * bestCompleteSolution.Cost
				{
					Q.push(withoutPivot);
				}
			}
		}
	}
}

fn main()
{
	let mut args : Vec<String> = env::args().collect();
	if args.len() < 2
	{
		//TODO: must output 36
		args.push("tests/asanov/input.txt".to_string());
	}
	
	let mut f = BufReader::new(File::open(args[1].to_string()).unwrap());
	let mut num_line = String::new();
	f.read_line(&mut num_line).unwrap();
	let n: usize = num_line.trim().parse().unwrap();
	
	let D: Vec<Vec<u32>> = f.lines()
		.take(n)
		.map(|l| l.unwrap().split(char::is_whitespace)
		.take(n)
		.map(|number| number.parse().unwrap())
		.collect())
	.collect();

	branch_and_bound(n, &D);
}